airbyte-integrations/connectors/source-shopify/source_shopify/utils.py

```diff
@@ -0,0 +1,93 @@
+#
+# MIT License
+#
+# Copyright (c) 2020 Airbyte
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in all
+# copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+#
+
+
+from functools import wraps
+from time import sleep
+
+import requests
+
+
+class ShopifyRequestRateLimits(object):
+
+    @staticmethod
+    def balance_rate_limit(threshold: float = 0.9, rate_limit_header: str = "X-Shopify-Shop-Api-Call-Limit"):
+        """
+        To avoid reaching Shopify API Rate Limits, use the "X-Shopify-Shop-Api-Call-Limit" header value,
+        to determine the current rate limits and load and handle wait_time based on load %.
+        Recomended wait_time between each request is 1 sec, we would handle this dynamicaly.
+
+        :: wait_time time between each request - 100 miliseconds
+
+        Header example:
+        {"X-Shopify-Shop-Api-Call-Limit": 10/40}, where: 10 - current load, 40 - max requests capacity.
+
+        More information: https://shopify.dev/api/usage/rate-limits
+        """
+
+        # Define standard timings in seconds
+        sleep_on_low_load: float = 0.2
+        sleep_on_avg_load: float = 1.5
+        sleep_on_high_load: float = 5.0
+        sleep_on_unknown_load: float = 1.0
```

Comment:
There is only 1 Shopify recommendation for general "less than rate-limit" data querying:
`2 requests/sec`...."Make requests with 1 sec delay". 

However this approach is a bit poor, because why should we wait 1 sec between each request if we have the capacity not to, and if we hit the approximate mid_load then we slow down and use some delay, once we have more capacity again - we use full burst. 
So, 
the 0.2 is the standard delay found experimentally - it's a perfect ratio between delay and request,
the 1.5 - is a compromise between recommended wait time and delay, where we wait 2 seconds to have another 4 requests capacity.
1.0 - is the standard recommended wait_time - we use it only if we don't have any `rate_limit` header.
5.0 - is the maximum time we wait in order to regain capacity back, each 1 sec of wait time gives us another 4 requests capacity, thus after 5 seconds of waiting without catch 429 HTTP Error gives us perfect rate_limit recovery and we have almost 90% of capacity back. We can always wait for 2.0 sec, as Retry-After provides us, but again, we have it only after 429 Error, and it's not sufficient, because should try to never catch that error.

Reply:
Thanks for detailed explanation, would be a nice to have that in the code as comment (or maybe the README.md?)